docs/yumdb.8
.\" yumdb command
.TH "yumdb" "8" "8 April 2010" "James Antill" ""
.SH "NAME"
yumdb \- query and alter the Yum database
.SH "SYNOPSIS"
\fByumdb\fP [command] [packages ...]
.SH "DESCRIPTION"
.PP 
This command is used to query and alter the yum database, which is a simple
key value store used in conjunction with the rpm database. Any installed package
can have arbitrary data in the yum database, however the main use case is to
store extra data about packages as they are installed.
.PP 
yumdb \fIcommand\fPs are:
.br 
.IP "\fByumdb get <key> [pkg-wildcard]...\fP"
.PP
This command will get the value for the given key, limiting to any specified
packages.
.br 
.IP "\fByumdb set <key> <value> [pkg-wildcard]..."
.PP
This command will set the value for the given key, to the given value, limiting
to any specified packages.
.br 
.IP "\fByumdb del <key> [pkg-wildcard]...
.PP
This command will delete the given key, limiting to any specified packages.
.br 
.IP "\fByumdb rename <old-key> <new-key> [pkg-wildcard]...
.PP
This command will rename the given old-key, to the given new-key, limiting to
any specified packages. If the old-key does not exist, nothing happens.
.br 
.IP "\fByumdb rename-force <old-key> <new-key> [pkg-wildcard]...
.PP
This command will rename the given old-key, to the given new-key, limiting to
any specified packages. If the old-key does not exist, new-key is deleted.
.br 
.IP "\fByumdb copy <old-key> <new-key> [pkg-wildcard]...
.PP
This command will copy the given old-key, to the given new-key, limiting to
any specified packages. If the old-key does not exist, nothing happens.
.br 
.IP "\fByumdb copy-force <old-key> <new-key> [pkg-wildcard]...
.PP
This command will copy the given old-key, to the given new-key, limiting to
any specified packages. If the old-key does not exist, new-key is deleted.
.br 
.IP "\fByumdb search <key> <wildcard>...
.PP
This command will search all packages for the given key, against any of the
given wildcard values.
.br 
.IP "\fByumdb exist <key> [pkg-wildcard]...
.PP
This command will print any packages which have the given key, limiting to any
specified packages.
.br 
.IP "\fByumdb unset <key> [pkg-wildcard]...
.PP
This command will print any packages which do not have the given key, limiting
to any specified packages.
.br 
.IP "\fByumdb info [pkg-wildcard]...
.PP
This command will display all the data stored in the yumdb, limiting to any
specified packages.
.br 
.IP "\fByumdb sync [pkg-wildcard]...
.PP
This command will add any missing data to the yumdb from the repositories,
limiting to any specified packages. This is useful to run if you have had any
aborted transactions (and thus. missing yumdb data).
Note that "yumdb sync" cannot know all the information that would have been put
into the yumdb at the time.
.br 
.IP "\fByumdb sync-force [pkg-wildcard]...
.PP
This command will replace any data in the yumdb from the repositories,
limiting to any specified packages.

.SH "EXAMPLES"
.PP
List all the packages which don't have a from_repo key/value:
.IP
yumdb unset from_repo
.PP
List all the packages which were installed as dependencies:
.IP
yumdb search reason dep

.SH "WELL KNOWN KEYS"
.PP
Note that there is no limit to the number of keys that can be created or what
they may contain (for installed packages only). However this is a list of well
known keys, and what they store.
.nf
.br 
.IP "\fBchecksum_data
.IP "\fBchecksum_type
.PP
These keys store the createrepo checksum, and it's type, of the available
package yum installed. Note that these are used by "yum version" to calculate
the rpmdb version.
.br 
.IP "\fBcommand_line
.PP
This key stores the entire command line, of the yum command (if it was called).
.br 
.IP "\fBfrom_repo
.IP "\fBfrom_repo_revision
.IP "\fBfrom_repo_timestamp
.PP
These keys take values from the available package yum installed, and store the
repo id, it's revision and timestamp.
.br 
.IP "\fBreason
.PP
This key stores either "user" or "dep", currently. To mark if the user requested
the package to be installed, or if it was brought in automatically as a
dependency. Note that this is kept over updates.
.br 
.IP "\fBreleasever
.PP
This key stores the value of releasever, when the package was installed.
.br 
.IP "\fBinstallonly
.PP
If this attribute has the value "keep" then this package will not be
removed automatically by the installonly process (and does not count towards
the installonly_limit).

.SH "SEE ALSO"
.nf
.I yum (8)
.I rpm (8)
.fi

.SH "AUTHORS"
.nf
James Antill <james.antill@redhat.com>.
.fi



ChangeLog
2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* yum-utils.spec: bumped yum-utils version to 1.1.31

2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* ChangeLog: updated ChangeLog

2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* yum-utils.spec: bumped yum-utils version to 1.1.31

2011-08-03  Shawn Starr <shawn.starr@rogers.com>

	* docs/yum-utils.1: Update yum-utils man page, to add new utils.

2011-07-25  James Antill <james@and.org>

	Merge branch 'master' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum-utils  * 'master' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum-utils: (3 commits)
	typo leaving .conf in ybc.conf.yumvar should just be ybc.yumvar
	...

2011-07-25  James Antill <james@and.org>

	* plugins/aliases/aliases: Add some new aliases, fix chkD.

2011-07-25  Seth Vidal <skvidal@fedoraproject.org>

	* yum-config-manager.py: typo leaving .conf in ybc.conf.yumvar
	should just be ybc.yumvar

2011-07-12  James Antill <james@and.org>

	* plugins/security/security.py: Also auto-apply --security when
	given --sec-severity for other commands.

2011-07-08  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Add repoquery @group completion.

2011-07-08  Ville Skyttä <ville.skytta@iki.fi>

	* repoquery.py: Fix "nitmes" typo in previous commit.

2011-07-06  James Antill <james@and.org>

	* repoquery.py: Fix typo

2011-06-29  James Antill <james@and.org>

	* repoquery.py: Allow @foo groups as a packages selector in
	repoquery.

2011-06-25  Ville Skyttä <ville.skytta@iki.fi>

	* plugins/aliases/aliases.py, plugins/security/security.py,
	plugins/versionlock/versionlock.py: Remove some unused local
	variables and imports flagged by pyflakes.

2011-06-24  Ville Skyttä <ville.skytta@iki.fi>

	* debuginfo-install.py, package-cleanup.py, yum-builddep.py,
	yum-complete-transaction.py: Check doUtilBuildTransaction return
	value, exit early on failure.  BZ 716267

2011-06-22  James Antill <james@and.org>

	* plugins/security/security.py: Try to be clever when sorting the
	security severities.

2011-06-17  Raman Gupta <raman@rocketraman.com>

	* plugins/merge-conf/merge-conf.py: After (external) merge is
	complete, give user option to delete other file

2011-06-17  Raman Gupta <raman@rocketraman.com>

	* plugins/merge-conf/merge-conf.py: Support meld as well as vimdiff
	for merging  When using meld, the left side has the "other file"
	i.e. the file with the .rpmnew or .rpmsave extension. The right side
	has the "final file" i.e. the target file into which changes have to
	be merged.  If the rpm action was to save the existing conf file,
	then the left side will be the original conf file (.rpmsave) and the
	right side will be the new rpm-based configuration. The user should
	merge any changes they want to keep from their old .rpmsave from
	left to right.  If the rpm action was to keep the existing conf
	file, then the left side will be the new conf file (.rpmnew) and
	right side will be the current local configuration. The user should
	merge any changes in the new rpm file into their local config from
	left to right.  Ideally these operations would be 3-way diffs
	between the local file and the new rpm config, using the original
	rpm config as the common merge base. However, the original rpm
	config is not available to yum and therefore a 3-way diff is not
	possible. Perhaps a future enhancement would be to use a .orig file,
	if present, as the merge base for meld.  When using vimdiff, the
	order has not changed. The first parameter to vimdiff is the "final:
	file i.e. the target file, and the second parameter is the "other"
	file i.e. the .rpmnew or .rpmsave file.

2011-06-10  James Antill <james@and.org>

	* repodiff.py: Huge speedup for large number of add/remove obsoletes
	processing.

2011-06-09  Nick Jacek <njacek@redhat.com>

	* yumdownloader.py: Stops yumdownloader from attempting to download
	the same package twice if it is available in multiple repos. BZ
	711767

2011-06-10  James Antill <james@and.org>

	* docs/repodiff.1, repodiff.py: Add --compare-arch for the other
	side of the multilib. diff. BZ 710579.

2011-06-09  Nick Jacek <njacek@redhat.com>

	* repodiff.py: Changed archlist set up to handle multilib. BZ 710579


2011-06-01  James Antill <james@and.org>

	* docs/repoquery.1: Add releasever and setopt docs. from yum to
	repoquery. BZ 638542.

2011-05-20  Seth Vidal <skvidal@fedoraproject.org>

	* yum-utils.spec: frelling pyyaml pkg named with weird case

2011-05-20  Seth Vidal <skvidal@fedoraproject.org>

	* plugins/puppetverify/puppetverify.conf,
	plugins/puppetverify/puppetverify.py, yum-utils.spec: add
	puppetverify plugin

2011-05-18  Frank Weltersbach <frankwe@live.com>

	* docs/package-cleanup.1: Fix typo in package-cleanup man page. BZ
	705617.

2011-05-17  Seth Vidal <skvidal@fedoraproject.org>

	* repoquery.py:     - fix rh bug: #705179     - fix fmt_queryformat not being happy
	    with new 'output' formats passed       everywhere else.

2011-04-26  James Antill <james@and.org>

	* yumdownloader.py: Allow files/provides to be used in
	yumdownloader.

2011-04-26  James Antill <james@and.org>

	* yumdownloader.py: Exit with code 2, if we can't download the
	packages asked for. BZ 699469.

2011-04-26  James Antill <james@and.org>

	* plugins/auto-update-debuginfo/auto-update-debuginfo.py: try/except
	the open, for the rpmdbv cached data (a-u-d plugin).

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Add yum-debug-dump completion.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-debug-dump.py: Mention filename arg in yum-debug-dump usage
	message.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-debug-dump.py: Make yum-debug-dump.py executable for running
	in-place.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Use functions from recent bash-completion if
	available.

2011-04-06  James Antill <james@and.org>

	* repoquery.py, yum-groups-manager.py: Convert

yumdb.py
#!/usr/bin/python -tt

import sys
import optparse
import fnmatch

import yum
import shlex

import os
import glob

parser = None

# FIXME: Internal knowledge
def _load_all_package_paths(db_path):
    # glob the path and get a dict of pkgs to their subdir
    glb = '%s/*/*/' % db_path
    pkgdirs = glob.glob(glb)
    _packages = {}
    for d in pkgdirs:
        if d[-1] == '/':
            d = d[:-1]
        pkgid = os.path.basename(d).split('-')[0]
        _packages[pkgid] = d
    return _packages

def setup_opts():
    version = "0.0.1"
    vers_txt = "Manage yum groups metadata version %s" % version
    usage_txt = """\
%prog <command> ...
      get           <key> [pkg-wildcard]...
      set           <key> <value> [pkg-wildcard]...
      del           <key> [pkg-wildcard]...
      rename        <key> <key> [pkg-wildcard]...
      rename-force  <key> <key> [pkg-wildcard]...
      copy          <key> <key> [pkg-wildcard]...
      search        <key> <wildcard>...
      exist?        <key> [pkg-wildcard]...
      unset?        <key> [pkg-wildcard]...
      info          [pkg-wildcard]...
      sync          [pkg-wildcard]...
      undeleted
      shell         [filename]...
"""
    parser =  optparse.OptionParser(usage = usage_txt, version = vers_txt)

    parser.add_option("--noplugins", action="store_false", default=True,
                      dest="plugins",
                      help="disable yum plugin support")
    parser.add_option("-c", "--config",
                      dest="conffile", help="config file location")

    return parser

def run_cmd(yb, args, inshell=False):
    if False: pass
    elif args[0] == 'get' and len(args) > 1:
        args.pop(0)
        ykey = args.pop(0)
        for pkg in sorted(yb.rpmdb.returnPackages(patterns=args)):
            print pkg
            if ykey in pkg.yumdb_info:
                print " " * 4, ykey, '=', getattr(pkg.yumdb_info, ykey)
            else:
                print " " * 4, ykey, '<unset>'
    elif args[0] == 'set' and len(args) > 2:
        args.pop(0)
        ykey = args.pop(0)
        yval = args.pop(0)
        for pkg in sorted(yb.rpmdb.returnPackages(patterns=args)):
            setattr(pkg.yumdb_info, ykey, yval)
            print pkg
            print " " * 4, ykey, '=', getattr(pkg.yumdb_info, ykey)
    elif args[0] in ('copy', 'copy-f', 'copy-force') and len(args) > 2:
        force = args[0] in ['copy-f', 'copy-force']
        args.pop(0)
        yokey = args.pop(0)
        ynkey = args.pop(0)
        for pkg in sorted(yb.rpmdb.returnPackages(patterns=args)):
            print pkg
            if yokey in pkg.yumdb_info:
                setattr(pkg.yumdb_info, ynkey, getattr(pkg.yumdb_info, yokey))
                print " " * 4, ynkey, '=', getattr(pkg.yumdb_info, ynkey)
            elif force and ynkey in pkg.yumdb_info:
                delattr(pkg.yumdb_info, ynkey)
                print " " * 4, ynkey, '<unset>'
            elif ynkey in pkg.yumdb_info:
                print " " * 4, ynkey, '=', getattr(pkg.yumdb_info, ynkey)
            else:
                print " " * 4, ynkey, '<unset>'
    elif args[0] in ['rename', 'rename-f', 'rename-force'] and len(args) > 2:
        force = args[0] in ['rename-f', 'rename-force']
        args.pop(0)
        yokey = args.pop(0)
        ynkey = args.pop(0)
        for pkg in sorted(yb.rpmdb.returnPackages(patterns=args)):
            print pkg
            if yokey in pkg.yumdb_info:
                setattr(pkg.yumdb_info, ynkey, getattr(pkg.yumdb_info, yokey))
                delattr(pkg.yumdb_info, yokey)
                print " " * 4, ynkey, '=', getattr(pkg.yumdb_info, ynkey)
            elif force and ynkey in pkg.yumdb_info:
                delattr(pkg.yumdb_info, ynkey)
                print " " * 4, ynkey, '<unset>'
            elif ynkey in pkg.yumdb_info:
                print " " * 4, ynkey, '=', getattr(pkg.yumdb_info, ynkey)
            else:
                print " " * 4, ynkey, '<unset>'
    elif args[0] in ['del', 'delete', 'rm', 'remove'] and len(args) > 1:
        args.pop(0)
        ykey = args.pop(0)
        for pkg in sorted(yb.rpmdb.returnPackages(patterns=args)):
            if ykey in pkg.yumdb_info:
                delattr(pkg.yumdb_info, ykey)
            print pkg
            print " " * 4, ykey, '<unset>'
    elif args[0] in ('search', 'search-q', 'search-quiet') and len(args) > 2:
        cmd = args.pop(0)
        ykey = args.pop(0)
        done = False
        # Maybe need some API so we don't have to load everything?
        for pkg in sorted(yb.rpmdb.returnPackages()):
            if ykey not in pkg.yumdb_info:
                continue
            found = False
            yval = getattr(pkg.yumdb_info, ykey)
            for arg in args:
                if fnmatch.fnmatch(yval, arg):
                    found = True
                    break
            if not found:
                continue
            if done and cmd == 'search': print ''
            done = True
            print pkg
            if cmd == 'search':
                print " " * 4, ykey, '=', yval
    elif args[0] in ['exist?', 'exist', 'exists'] and len(args) > 1:
        args.pop(0)
        ykey = args.pop(0)
        for pkg in sorted(yb.rpmdb.returnPackages(patterns=args)):
            if ykey not in pkg.yumdb_info:
                continue
            print pkg
    elif args[0] in ['unset?', 'unset'] and len(args) > 1:
        args.pop(0)
        ykey = args.pop(0)
        for pkg in sorted(yb.rpmdb.returnPackages(patterns=args)):
            if ykey in pkg.yumdb_info:
                continue
            print pkg
    elif args[0] == 'info':
        args.pop(0)
        done = False
        for pkg in sorted(yb.rpmdb.returnPackages(patterns=args)):
            if done: print ''
            done = True
            print pkg
            for ykey in sorted(pkg.yumdb_info):
                print " " * 4, ykey, '=', getattr(pkg.yumdb_info, ykey)
    elif args[0] in ('sync', 'synchronize', 'sync-f', 'synchronize-f',
                     'sync-force', 'synchronize-force'):
        force = args[0] in ('sync-f', 'synchronize-f',
                            'sync-force', 'synchronize-force')
        args.pop(0)
        done = False
        for pkg in sorted(yb.rpmdb.returnPackages(patterns=args)):
            apkg = yb.pkgSack.searchPkgTuple(pkg.pkgtup)
            if not apkg:
                continue
            apkg = sorted(apkg)[0]

            if done: print ''
            done = True
            print pkg
            ndata = {}
            ndata['releasever'] = yb.conf.yumvar['releasever']
            ndata['from_repo']  = apkg.repoid
            csum = apkg.returnIdSum()
            if csum is not None:
                ndata['checksum_type'] = str(csum[0])
                ndata['checksum_data'] = str(csum[1])
            if hasattr(apkg.repo, 'repoXML'):
                md = apkg.repo.repoXML
                if md and md.revision is not None:
                    ndata['from_repo_revision']  = str(md.revision)
                if md:
                    ndata['from_repo_timestamp'] = str(md.timestamp)

            loginuid = yum.misc.getloginuid()
            if loginuid is not None:
                ndata['changed_by'] = str(loginuid)

            for ykey in ndata:
                if not force and hasattr(pkg.yumdb_info, ykey):
                    continue
                setattr(pkg.yumdb_info, ykey, ndata[ykey])
            for ykey in sorted(pkg.yumdb_info):
                if ykey not in ndata:
                    continue
                print " " * 4, ykey, '=', getattr(pkg.yumdb_info, ykey)
    elif args[0] == 'undeleted':
        yumdb_packages = _load_all_package_paths(yb.rpmdb.yumdb.conf.db_path)
        for pkg in sorted(yb.rpmdb.returnPackages()):
            if pkg.pkgid in yumdb_packages:
                del yumdb_packages[pkg.pkgid]
        for pkgid in yumdb_packages:
            print "%s: %s" % (pkgid, yumdb_packages[pkgid])

    elif args[0] == 'shell' and not inshell:
        args.pop(0)
        if args:
            fos = []
            for arg in args:
                fos.append(open(arg))
        else:
            fos = [sys.stdin]
        for fo in fos:
            print "=" * 79
            for line in fo:
                run_cmd(yb, shlex.split(line), inshell=True)
                print "-" * 79
    else:
        print >>sys.stderr, parser.format_help()
        sys.exit(1)

def main():
    global parser

    parser = setup_opts()
    (opts, args) = parser.parse_args()

    yb = yum.YumBase()
    if opts.conffile:
        yb.preconf.fn = opts.conffile
    if not opts.plugins:
        yb.preconf.init_plugins = False
    yb.conf

    if len(args) < 1:
        print >>sys.stderr, parser.format_help()
        sys.exit(1)

    run_cmd(yb, args)


if __name__ == '__main__':
    main()
